Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md

```diff
@@ -523,6 +523,93 @@ A similar approach works. Answer the following questions:
+On iOS the `proc` directory is not available, but you can list the dynamic libraries in an app with the function `_dyld_image_count`.
+
+The other method (which also works for non-jailbroken devices) consists of embedding a [frida-gadget](https://www.frida.re/docs/gadget/ "Frida Gadget") into the IPA and _forcing_ the app to load it as one of its native libraries.
+
+In the application.app data directory the app stores the static content, as well as the application's ARM-compiled binary including the apps's libaries. If you inspect the app's directory in `/var/containers/Bundle/Application/[UUID]/[Application].app` you'll find the embedded frida-gadget as FridaGadget.dylib.
+
+```bash
+Svens-iPhone:/var/containers/Bundle/Application/AC5DC1FD-3420-42F3-8CB5-E9D77C4B287A/SwiftSecurity.app/Frameworks root# ls -alh
+total 87M
+drwxr-xr-x 10 _installd _installd  320 Nov 19 06:08 ./
+drwxr-xr-x 11 _installd _installd  352 Nov 19 06:08 ../
+-rw-r--r--  1 _installd _installd  70M Nov 16 06:37 FridaGadget.dylib
+-rw-r--r--  1 _installd _installd 3.8M Nov 16 06:37 libswiftCore.dylib
+-rw-r--r--  1 _installd _installd  71K Nov 16 06:37 libswiftCoreFoundation.dylib
+-rw-r--r--  1 _installd _installd 136K Nov 16 06:38 libswiftCoreGraphics.dylib
+-rw-r--r--  1 _installd _installd  99K Nov 16 06:37 libswiftDarwin.dylib
+-rw-r--r--  1 _installd _installd 189K Nov 16 06:37 libswiftDispatch.dylib
+-rw-r--r--  1 _installd _installd 1.9M Nov 16 06:38 libswiftFoundation.dylib
+-rw-r--r--  1 _installd _installd  76K Nov 16 06:37 libswiftObjectiveC.dylib
+```
+
+Looking at these _traces_ that Frida _lefts behind_, you might already imagine that detecting those would be a trivial task. And actually, so trivial will be bypassing that detection. But things can get much more complicated. The following table shortly presents a set of some typical Frida detection methods and a short discussion on their effectiveness.
+
+> Some of the following detection methods are implemented in the [iOS Security Suite](https://github.com/securing/IOSSecuritySuite "iOS Security Suite").
+
+| Method | Description | Discussion |
+| --- | --- | --- |
+| **Check Signature of an App** | One detection mechanism would be to verify the signature of the app. In order to embed the frida-gadget within the IPA, it would need to be repackaged and resigned. There is a library available in Objective-C called [iOS-App-Security-Class](https://github.com/karek314/iOS-App-Security-Class "iOS-App-Security-Class") that checks if the app is encrypted or not, by verifying `cryptid`. If this value is 0, the app was dumped (e.g. "clutched") and the Fairplay DRM encryption is not present anymore.  | This indicates that someone is trying to tamper and reverse engineer an app. There are limitations in such an implementation, as you cannot easily test if the check for `cryptid` is properly working. The app will only be encrypted by Apple after you submitted it to the App Store. |
```

Comment:
```suggestion
| **Check Signature of an App** | One detection mechanism would be to verify the signature of the app. In order to embed the frida-gadget within the IPA, it would need to be repackaged and resigned. There is a library available in Objective-C called [iOS-App-Security-Class](https://github.com/karek314/iOS-App-Security-Class "iOS-App-Security-Class") that checks if the app is encrypted or not, by verifying `cryptid`. If this value is 0, the app was dumped (e.g. "clutched") and the Fairplay DRM encryption is not present anymore.  | This indicates that someone is trying to tamper and reverse engineer the app. There are limitations in such an implementation, as you cannot easily test if the check for `cryptid` is properly working.  |
```
I removed that last part, as I think it's unnecessary. With enterprise certs, too, the app would be signed, right? So when is that ever an issue when preventing RE? (I can only think of "hard to test, since it will complain during dev", or maybe even when submitting it to the app store?) In any case, that line could use some more explanation.

Reply:
I was going through the commit history of the iOS-App-Security-Class and the last change to the project is 1 year ago, but the last code commit to the library is already over 2 years old and there is an open issue that was not addressed and fixed for almost a year now. This library is not maintained. 
I removed the whole row now, as there is no API offered by the SDK to validate the signature on iOS. Suggesting this "hack" is not a good control and you are right it's also hard to develop and I am not sure what happens during the app review process when using it.

Reply:
Thanks for pointing out